### PR TITLE
drivers: sensors: adt7420: Add driver for ADT7420 Temperature Sensor

### PR DESCRIPTION
**Pull request by mhennerich**
This patch adds support for the Analog Devices ADT7420 High-Accuracy
16-bit Digital I2C Temperature Sensors. Optionally sensor threshold
events are supported.

Signed-off-by: Michael Hennerich <michael.hennerich@analog.com>

### CODE REVIEW

File: drivers/sensor/adt7420/adt7420.c [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+			return -EIO;
+		}
+
+		return 0;
+	case SENSOR_ATTR_UPPER_THRESH:
+		reg = ADT7420_REG_T_HIGH_MSB;
+		/* Fallthrough */
+	case SENSOR_ATTR_LOWER_THRESH:
+		if (!reg)
+			reg = ADT7420_REG_T_LOW_MSB;
```
Comment (MaureenHelm) [inline drivers/sensor/adt7420/adt7420.c]: Please wrap in braces

File: drivers/sensor/adt7420/adt7420.c [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+
+		value = (s64_t)val->val1 * 1000000 + val->val2;
+		value = (value / ADT7420_TEMP_SCALE) << 1;
```
Comment (MaureenHelm) [inline drivers/sensor/adt7420/adt7420.c]: Is it possible this result could be out of range before you cast it to s16_t and pass to `adt7420_temp_reg_write`?

File: drivers/sensor/adt7420/adt7420.c [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+static int adt7420_probe(struct device *dev)
+{
+	struct adt7420_data *drv_data = dev->driver_data;
+	u8_t value;
+	int ret;
+
+	ret = i2c_reg_read_byte(drv_data->i2c, drv_data->i2c_addr,
+				ADT7420_REG_ID, &value);
+	if (ret)
+		return ret;
```
Comment (MaureenHelm) [inline drivers/sensor/adt7420/adt7420.c]: Please add braces after all the if statements.
http://docs.zephyrproject.org/contribute/contribute_guidelines.html#coding-style

File: drivers/sensor/adt7420/adt7420.c [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+	if (adt7420_init_interrupt(dev) < 0) {
+		SYS_LOG_ERR("Failed to initialize interrupt!");
+		return -EIO;
+	}
+#endif
+
+	return 0;
+}
+
+int adt7420_init(struct device *dev)
```
Comment (MaureenHelm) [inline drivers/sensor/adt7420/adt7420.c]: This can be static like the other functions.

File: drivers/sensor/adt7420/adt7420.c [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+	struct adt7420_data *drv_data = dev->driver_data;
+
+	drv_data->i2c = device_get_binding(CONFIG_ADT7420_I2C_MASTER_DEV_NAME);
+	if (drv_data->i2c == NULL) {
+		SYS_LOG_DBG("Failed to get pointer to %s device!",
+			    CONFIG_ADT7420_I2C_MASTER_DEV_NAME);
+		return -EINVAL;
+	}
+
+	drv_data->i2c_addr = CONFIG_ADT7420_I2C_ADDR;
```
Comment (MaureenHelm) [inline drivers/sensor/adt7420/adt7420.c]: Move the i2c addr to a const config structure and remove this line. You can also add the i2c_dev name to the const config structure instead of passing `CONFIG_ADT7420_I2C_MASTER_DEV_NAME` directly to `device_get_binding`

File: drivers/sensor/adt7420/adt7420_trigger.c [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+
+#include <device.h>
+#include <gpio.h>
+#include <i2c.h>
+#include <misc/util.h>
+#include <kernel.h>
+#include <sensor.h>
+#include "adt7420.h"
+
+extern struct adt7420_data adt7420_driver;
```
Comment (MaureenHelm) [inline drivers/sensor/adt7420/adt7420_trigger.c]: Why is this here?

File: drivers/sensor/adt7420/adt7420_trigger.c [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+	gpio_pin_enable_callback(drv_data->gpio, CONFIG_ADT7420_GPIO_PIN_NUM);
+
+	return 0;
+}
+
+int adt7420_init_interrupt(struct device *dev)
+{
+	struct adt7420_data *drv_data = dev->driver_data;
+
+	drv_data->gpio = device_get_binding(CONFIG_ADT7420_GPIO_DEV_NAME);
```
Comment (MaureenHelm) [inline drivers/sensor/adt7420/adt7420_trigger.c]: `CONFIG_ADT7420_GPIO_DEV_NAME` and `CONFIG_ADT7420_GPIO_PIN_NUM` can also be stored in the const config structure

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Please add a sample or update tests/drivers/build_all/sensors_a_m.conf to include this driver. This will ensure the driver gets built in CI. Are there any in-tree boards that have this sensor?

File: drivers/sensor/adt7420/Kconfig [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+
+menuconfig ADT7420
+	bool
+	prompt "ADT7420 Temperature Sensor"
+	depends on I2C
+	help
+	  Enable the driver for Analog Devices ADT7420 High-Accuracy
+	  16-bit Digital I2C Temperature Sensors.
+
+config ADT7420_NAME
```
[the remark this answers is not in the document]
Comment (mhennerich) [reply drivers/sensor/adt7420/Kconfig]: Hi Maureen,

Thanks for your feedback!

In theory device tree is used to describe bus hierarchies and devices that can appear under a bus. 
Basically it describes the physical devices in a system, which cannot be automatically probed.

I totally agree that the parent bus and slave addresses, etc. should come from the device tree.
Practically this would come from the parent device which will instantiate all its devices, via a device specific structure.
In fact that’s the only way to allow instantiating the same driver multiple times…
Which I have no idea, how that can be done in Zephyr today.
 
This being said,  Zephyr device tree usage is very limited. The user must provide a fixup file which maps
device tree properties to CONFIG options.

So my question is what do you mean by “should come from the device tree” ?
Isn’t what’s implemented in this driver exactly the same what all other drivers/sensor do?

File: drivers/sensor/adt7420/Kconfig [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+	string
+	prompt "Driver name"
+	default "ADT7420"
+	depends on ADT7420
```
[the remark this answers is not in the document]
Comment (mhennerich) [reply drivers/sensor/adt7420/Kconfig]: Will do

File: drivers/sensor/adt7420/Kconfig [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+	default 5
+	help
+	  Specify the temperature hysteresis in °C for the THIGH, TLOW,
+	  and TCRIT temperature limits.
+
+config ADT7420_TEMP_CRIT
+	int "Critical overtemperature in °C"
+	depends on ADT7420
+	range -40 150
+	default 147
```
[the remark this answers is not in the document]
Comment (mhennerich) [reply drivers/sensor/adt7420/Kconfig]: Yes 147 °C is the critical over temperature reset default of the ADT7420. So if no one changes this config, it'll use the device default.

File: drivers/sensor/adt7420/Kconfig [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+	depends on GPIO
+	select ADT7420_TRIGGER
+
+endchoice
+
+config ADT7420_TRIGGER
+	bool
+	depends on ADT7420
+
+config ADT7420_GPIO_DEV_NAME
```
[the remark this answers is not in the document]
Comment (mhennerich) [reply drivers/sensor/adt7420/Kconfig]: Please see previous question

File: drivers/sensor/adt7420/adt7420.c (hunk 1 above)
Comment (mhennerich) [reply drivers/sensor/adt7420/adt7420.c]: Please bear with me, that's my first Zephyr patch. I thought that we follow Linux kernel coding style.
Braces around single line code blocks seems to be an exception...

File: drivers/sensor/adt7420/adt7420.c (hunk 2 above)
Comment (mhennerich) [reply drivers/sensor/adt7420/adt7420.c]: No - not if you pass valid threshold values.

File: drivers/sensor/adt7420/adt7420.c (hunk 3 above)
Comment (mhennerich) [reply drivers/sensor/adt7420/adt7420.c]: sure

File: drivers/sensor/adt7420/adt7420.c (hunk 4 above)
Comment (mhennerich) [reply drivers/sensor/adt7420/adt7420.c]: good catch

File: drivers/sensor/adt7420/adt7420_trigger.c (hunk 6 above)
Comment (mhennerich) [reply drivers/sensor/adt7420/adt7420_trigger.c]: not needed at all

File: drivers/sensor/adt7420/adt7420.c (hunk 5 above)
Comment (mhennerich) [reply drivers/sensor/adt7420/adt7420.c]: Thanks. Makes sense to keep the config stuff in one place.

File: drivers/sensor/adt7420/adt7420_trigger.c (hunk 7 above)
Comment (mhennerich) [reply drivers/sensor/adt7420/adt7420_trigger.c]: will do.

**Comment by mhennerich** [comment]
currently there are no in-tree boards with this sensor.
Shall I merge my samples/sensor/adt7420 example into this PR or shall I open a separate one?

Thanks,
Michael

File: drivers/sensor/adt7420/Kconfig (hunk 8 above)
[the remark this answers is not in the document]
Comment (MaureenHelm) [reply drivers/sensor/adt7420/Kconfig]: It's been a long process converting drivers over to device tree, so you will find that some drivers are further along than others. For new drivers, however, we do want to enforce getting hardware information from device tree from the start. Please have a look at `drivers/sensors/fxos8700` and `boards/arm/frdm_k64f/frdm_k64f.dts` for an example of what we're looking for.

File: drivers/sensor/adt7420/Kconfig (hunk 10 above)
[the remark this answers is not in the document]
Comment (MaureenHelm) [reply drivers/sensor/adt7420/Kconfig]: Ok, cool. Will you please note this in the help?

File: drivers/sensor/adt7420/adt7420.c (hunk 1 above)
Comment (MaureenHelm) [reply drivers/sensor/adt7420/adt7420.c]: Yes, this is an exception to the Linux coding style.

File: drivers/sensor/adt7420/Kconfig (hunk 8 above)
[the remark this answers is not in the document]
Comment (mhennerich) [reply drivers/sensor/adt7420/Kconfig]: Oh I see - you asking me to guard these config options with "if !HAS_DTS_I2C_DEVICE" !?

File: drivers/sensor/adt7420/adt7420.c (hunk 2 above)
Comment (MaureenHelm) [reply drivers/sensor/adt7420/adt7420.c]: Then you should add a check to ensure the threshold values are valid

**Comment by MaureenHelm** [comment]
> Shall I merge my samples/sensor/adt7420 example into this PR or shall I open a separate one?

Please add it in another commit in this PR.

File: drivers/sensor/adt7420/Kconfig (hunk 8 above)
[the remark this answers is not in the document]
Comment (MaureenHelm) [reply drivers/sensor/adt7420/Kconfig]: Yes, and to add a binding in `dts/bindings/sensor/`

**Comment by mhennerich** [comment]
Hi,

I think I addressed all the feedback and issues.
However CI fails now on some platforms which I guess using HAS_DTS_I2C_DEVICE.

/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/drivers/sensor/adt7420/adt7420.c:214:14: error: ‘CONFIG_ADT7420_I2C_MASTER_DEV_NAME’ undeclared here (not in a function)

Wondering how shall we avoid these expected failing CI tests?

-Michael

**Comment by mhennerich** [comment]
Ok never mind - I fixed (and pushed) the build issue on boards using CONFIG_HAS_DTS_I2C_DEVICE, by providing  a dummy dts.fixup in case some of the required CONFIG_ADT7240_ defines are not present in the device tree.

File: drivers/sensor/adt7420/Kconfig [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+	bool "Use own thread"
+	depends on GPIO
+	select ADT7420_TRIGGER
+
+endchoice
+
+config ADT7420_TRIGGER
+	bool
+
+if !HAS_DTS_I2C_DEVICE
```
Comment (MaureenHelm) [inline drivers/sensor/adt7420/Kconfig]: This should be `if !HAS_DTS_GPIO_DEVICE`

File: samples/sensor/adt7420/src/main.c [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+		printf("temperature %.6f C\n",
+		       sensor_value_to_double(&temp_val));
+
+		k_sleep(1000);
+	}
+}
+
+void main(void)
+{
+	struct device *dev = device_get_binding("ADT7420");
```
Comment (MaureenHelm) [inline samples/sensor/adt7420/src/main.c]: Use `CONFIG_ADT7420_NAME` instead of the hardcoded string

File: samples/sensor/adt7420/src/main.c [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+					 &temp_val);
+		if (ret) {
+			printf("sensor_channel_get failed ret %d\n", ret);
+			return;
+		}
+
+		printf("temperature %.6f C\n",
+		       sensor_value_to_double(&temp_val));
+
+		k_sleep(1000);
```
Comment (MaureenHelm) [inline samples/sensor/adt7420/src/main.c]: You probably don't want this if trigger is enabled

File: samples/sensor/adt7420/README.txt [hunk 15]
```diff
@@ -0,0 +1,5 @@
+Title: adt7420
+
+Description:
+
+A example using the ADT7420 high accuracy digital temperature sensor.
```
Comment (MaureenHelm) [inline samples/sensor/adt7420/README.txt]: I expect @dbkinder will ask for more details, like how to build the sample and what is the expected output.

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: drivers/sensor/adt7420/Kconfig (hunk 12 above)
Comment (mhennerich) [reply drivers/sensor/adt7420/Kconfig]: right

File: samples/sensor/adt7420/src/main.c (hunk 14 above)
Comment (mhennerich) [reply samples/sensor/adt7420/src/main.c]: Yep

File: samples/sensor/adt7420/README.txt (hunk 15 above)
Comment (mhennerich) [reply samples/sensor/adt7420/README.txt]: Just looking at the other examples - I'll do the whole thing including reference, wiring building and running. Thanks for your review. I'll push the fixes right now.

**Review by MaureenHelm — APPROVED** [review]

File: samples/sensor/adt7420/README.rst [hunk 16]
```diff
@@ -0,0 +1,55 @@
+.. _adt7420:
+
+ADT7420: High accuracy digital I2C temperature sensor
+#####################################################
+
+Desciption
```
Comment (dbkinder) [inline samples/sensor/adt7420/README.rst]: ``Description``

File: samples/sensor/adt7420/README.rst [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+**********
+
+This sample application periodically (1Hz) measures the ambient temperature
+in degrees Celsius. The result is written to the console.
+Optionally it also shows how to use the upper and lower threshold triggers.
```
Comment (dbkinder) [inline samples/sensor/adt7420/README.rst]: Add a comma, ``Optionally, it``

File: samples/sensor/adt7420/README.rst [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+**********
+
+ - ADT7420: http://www.analog.com/adt7420
+
+Wiring
+*******
+
+This sample uses the ADT7420 sensor controlled using the I2C interface.
+Connect Supply: **VDD**, **GND** and Interface: **SDA**, **SCL**
+optionally connect the **INT** to a interrupt capabale GPIO.
```
Comment (dbkinder) [inline samples/sensor/adt7420/README.rst]: add ``and optionally``

``capable``

**Review by dbkinder — DISMISSED** [review]
pretty close on documentation... a few misspellings and punctuation fixes.  Thanks!

File: samples/sensor/adt7420/README.rst (hunk 16 above)
Comment (mhennerich) [reply samples/sensor/adt7420/README.rst]: Good catch

File: samples/sensor/adt7420/README.rst (hunk 17 above)
Comment (mhennerich) [reply samples/sensor/adt7420/README.rst]: ok

File: samples/sensor/adt7420/README.rst (hunk 18 above)
Comment (mhennerich) [reply samples/sensor/adt7420/README.rst]: Will do. Thanks for the Review!
I already fixed and pushed them.
